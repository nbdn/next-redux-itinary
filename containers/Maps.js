/*global google:true*/
/*eslint no-undef: "error"*/
import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';

import GoogleMapsComponent from '../components/GoogleMaps';

import { getOrderedPlaces } from '../redux/maps';

import Page from '../components/Page';

class MapsContainer extends PureComponent {
  state = {};

  componentWillReceiveProps({ places }) {
    if (Array.isArray(places) && places.length >= 2) {
      this.computeDirections(places);
    }
  }

  computeDirections(places) {
    const DirectionsService = new google.maps.DirectionsService();
    const start = places.shift();
    const end = places.shift();
    const waypoints = places.map(p => {
      return { location: p.address, stopover: true };
    });
    DirectionsService.route(
      {
        origin: start.address,
        destination: end.address,
        travelMode: google.maps.TravelMode.DRIVING,
        waypoints
      },
      (result, status) => {
        if (status === google.maps.DirectionsStatus.OK) {
          this.setState({
            directions: result
          });
        } else {
          /* eslint-disable no-console */
          console.error(`error fetching directions ${result}`);
          /* eslint-enable */
        }
      }
    );
  }

  render() {
    const { places } = this.props;
    const { directions } = this.state;

    return (
      <GoogleMapsComponent
        directions={directions}
        places={directions ? [] : places}
      />
    );
  }
}

MapsContainer.propTypes = {
  places: PropTypes.array
};

const mapStateToProps = createStructuredSelector({
  places: getOrderedPlaces
});

export default Page(connect(mapStateToProps)(MapsContainer));
