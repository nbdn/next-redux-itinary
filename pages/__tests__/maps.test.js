/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';

import Maps from '../maps.js';

describe('Snapshot Testing', () => {
  it('Maps mount without crashing', () => {
    const mockUrl = { query: '' };
    const component = renderer.create(<Maps url={mockUrl} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
