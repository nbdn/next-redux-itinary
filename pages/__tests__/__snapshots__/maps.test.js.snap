// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing Maps mount without crashing 1`] = `
<div
  style={
    Object {
      "height": "100%",
    }
  }
/>
`;
